### PR TITLE
Improvement/streamline jquery plugins

### LINKED ISSUE #468: jQuery plugins don't have destruction/cleanup functionality
I'm currently build a React Component to for dropdown, to utilize our jQuery plugin.
Construction is no problem, but as soon as my component unmount I need to cleanup the markup which is not controlled by React directly.

In case of dropdown, the bound events should be properly unbound.

### LINKED ISSUE #537: streamline jQuery-Plugin Boilerplate and document it
At the moment we use sightly different glue-code for our custom jQuery-Plugins.
Plus instantiation of those happen after `onload` but should be done at `DOMready` for performance reasons.

Furthermore every jQuery-Plugin stores a reference of itself to the data warehouse of the associated jQuery collection, plus HTML5 Data-Attributes are utilised to assign options.

All of this should be streamlined and documented. Ideally we have a little plugin module which does all the magic and can just be extended with a specific plugin.

To support options namespacing we could consider [$.html5data](http://markdalgleish.com/projects/jquery-html5data/)

Feature Request:

- [ ] Bulletproof against multiple instantiations
- [ ] Clean-up
- [ ] Options through HTML5 data attributes
 - [ ] make sure to handle namespace issues properly
 - [ ] allow for global defaults

Ressources:

- **Official**: [How to Create a Basic Plugin](http://learn.jquery.com/plugins/basic-plugin-creation/)
- **Official**: [Advanced Plugin Concepts](http://learn.jquery.com/plugins/advanced-plugin-concepts/)
- [jQuery Boilerplate](https://jqueryboilerplate.com/)
- [Extending jQuery Boilerplate](https://github.com/jquery-boilerplate/jquery-boilerplate/wiki/Extending-jQuery-Boilerplate)
- [Handling plugin defaults and predefinitions](https://github.com/jquery-boilerplate/jquery-boilerplate/wiki/Handling-plugin-defaults-and-predefinitions)
- [highly-configurable.plugin.boilerplate](https://github.com/jquery-boilerplate/jquery-patterns/blob/master/patterns/jquery.highly-configurable.plugin.boilerplate.js)
- [jquery.boilerplate.js](https://github.com/jquery-boilerplate/jquery-boilerplate/blob/master/src/jquery.boilerplate.js)
- [jQuery Plugin Patterns](https://github.com/jquery-boilerplate/jquery-patterns)

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #537 fixes #468 

At the moment we use sightly different glue-code for our custom jQuery-Plugins.
Plus instantiation of those happen after `onload` but should be done at `DOMready` for performance reasons.

Furthermore every jQuery-Plugin stores a reference of itself to the data warehouse of the associated jQuery collection, plus HTML5 Data-Attributes are utilised to assign options.

All of this should be streamlined and documented. Ideally we have a little plugin module which does all the magic and can just be extended with a specific plugin.

To support options namespacing we could consider [$.html5data](http://markdalgleish.com/projects/jquery-html5data/)

Feature Request:

- [x] Bulletproof against multiple instantiations
- [x] Support optional custom instantiation callback
- [x] Support various plugins being attached to a single DOM node (avoid namespace issues)
- [x] Basic clean-up
- [x] Options through HTML5 data attributes
 - [x] make sure to handle namespace issues properly
 - [x] namespaced options overwrite non-namespaced options
 - [x] allow for global defaults
- [x] add proper inline jsDoc doclets
- [x] add proper documention to style-guide
- [x] utilize Plugin Boilerplate glue code for all custom jQuery plugins

Ressources:

- **Official**: [How to Create a Basic Plugin](http://learn.jquery.com/plugins/basic-plugin-creation/)
- **Official**: [Advanced Plugin Concepts](http://learn.jquery.com/plugins/advanced-plugin-concepts/)
- [jQuery Boilerplate](https://jqueryboilerplate.com/)
- [Extending jQuery Boilerplate](https://github.com/jquery-boilerplate/jquery-boilerplate/wiki/Extending-jQuery-Boilerplate)
- [Handling plugin defaults and predefinitions](https://github.com/jquery-boilerplate/jquery-boilerplate/wiki/Handling-plugin-defaults-and-predefinitions)
- [highly-configurable.plugin.boilerplate](https://github.com/jquery-boilerplate/jquery-patterns/blob/master/patterns/jquery.highly-configurable.plugin.boilerplate.js)
- [jquery.boilerplate.js](https://github.com/jquery-boilerplate/jquery-boilerplate/blob/master/src/jquery.boilerplate.js)
- [jQuery Plugin Patterns](https://github.com/jquery-boilerplate/jquery-patterns)

### CODE REVIEW

**Comment by AndyOGo** [comment]
I'm almost done.
One big think is left. I want the API documentation to be automatically generated and handed over to metalsmith to display.
I think we have basically two options:

- [jsdoc-to-markdown](https://github.com/jsdoc2md/jsdoc-to-markdown)
- [jsdoc-parse](https://github.com/jsdoc2md/jsdoc-parse)

**Comment by AndyOGo** [comment]
Regarding documentation of jQuery plugins.
I found just this page: https://design.axa.ch/fundamentals/code/getting-started.html#jquery-plugins

I would say we need a basic jQuery plugins page which documentation how our jQuery plugins are supposed to work and a list of all available plugins with links to their automatically generated docs (source from jsdoc)

**Comment by AndyOGo** [comment]
We could utilize gulp-inject for the injection process too.

**Comment by AndyOGo** [comment]
Okay I'm done.
I decided for the quick manual inclusion of jsdoc markdown files.

**Comment by AndyOGo** [comment]
I just found that the menu plugins (`menu-collapsing`, `menu-main`, `menu-sliding`) are setting the value of the `data-menu` attribute to instantiate the associated plugin. See current code

```js
$('[data-menu="collapsing"]')
```

```js
$('[data-menu="main"]')
```

```js
$('[data-menu="sliding"]')
```

**Comment by AndyOGo** [comment]
@axa-ch/web-style-guide 
I'm almost done, one thing is missing testing, code-review.
Docs are written and I updated all existing jQuery plugins.

**Comment by elessar-ch** [comment]
Looks really great.
How should we present them in the docs? I think it's might be best to have a section under components like we have for "Form and Input Elements".

**Comment by AndyOGo** [comment]
:) Yeah, I'm curios about that.
So far I added one basic doc for the boilerplate code at https://design.axa.ch/fundamentals/code/jquery-plugins.html

For each jQuery plugin I though of:
* creating a list of all plugins
* including each generated markdown file at the corresponding component

File: jquery/plugin.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+ *  which: false
+ * })
+ *
+ * @example <caption>Calling a method of the plugin</caption>
+ * var locked = $('#lock-me').lockDimension()
+ *
+ * // call public method
+ * locked.lockDimension('unlock')
+ */
+function Plugin(name, Constructor, customInstantiationCB) {
```
Comment (sventschui) [inline jquery/plugin.js]: Lower case function name.
Wouldn't registerPlugin aussagekräftiger sein?

Comment (AndyOGo) [inline jquery/plugin.js]: @sventschui truely true.
Will update

**Comment by AndyOGo** [comment]
I quickly tested all component pages.

`checkbox-group` long does not work, though the smaller example is working. weird.
Could not find the cause so far...

**Comment by AndyOGo** [comment]
I think i found it, the original `checkbox` plugin trigger onload not on DOM-ready...

**Comment by AndyOGo** [comment]
@sventschui @elessar-ch 
I quickly tested all components of the style-guide and fixed those which I found to be broken.
Plugin is properly renamed to registerPlugin.

I would be ready to merge 😄

**Comment by sventschui** [comment]
Awesome!

**Pull request merged**